File: FlowCrypt/src/main/java/com/flowcrypt/email/security/SecurityUtils.java
/*
 * © 2016-2019 FlowCrypt Limited. Limitations apply. Contact human@flowcrypt.com
 * Contributors: DenBond7
 */

package com.flowcrypt.email.security;

import android.content.Context;
import android.database.Cursor;
import android.text.TextUtils;

import com.flowcrypt.email.Constants;
import com.flowcrypt.email.api.retrofit.node.NodeCallsExecutor;
import com.flowcrypt.email.api.retrofit.response.model.node.NodeKeyDetails;
import com.flowcrypt.email.api.retrofit.response.node.EncryptKeyResult;
import com.flowcrypt.email.api.retrofit.response.node.ZxcvbnStrengthBarResult;
import com.flowcrypt.email.database.dao.source.AccountDao;
import com.flowcrypt.email.database.dao.source.ContactsDaoSource;
import com.flowcrypt.email.database.dao.source.KeysDaoSource;
import com.flowcrypt.email.database.dao.source.UserIdEmailsKeysDaoSource;
import com.flowcrypt.email.model.PgpContact;
import com.flowcrypt.email.model.PgpKeyInfo;
import com.flowcrypt.email.security.model.PrivateKeyInfo;
import com.flowcrypt.email.util.exception.DifferentPassPhrasesException;
import com.flowcrypt.email.util.exception.NoKeyAvailableException;
import com.flowcrypt.email.util.exception.NoPrivateKeysAvailableException;
import com.flowcrypt.email.util.exception.NodeException;
import com.flowcrypt.email.util.exception.PrivateKeyStrengthException;
import com.google.android.gms.common.util.CollectionUtils;
import com.nulabinc.zxcvbn.Zxcvbn;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * This class help to receive security information.
 *
 * @author DenBond7
 * Date: 05.05.2017
 * Time: 13:08
 * E-mail: DenBond7@gmail.com
 */

public class SecurityUtils {

  /**
   * Get a PrivateKeyInfo list.
   *
   * @param context Interface to global information about an application environment.
   * @return <tt>List<PrivateKeyInfo></tt> Return a list of PrivateKeyInfo objects.
   */
  public static List<PrivateKeyInfo> getPrivateKeysInfo(Context context) throws Exception {
    ArrayList<PrivateKeyInfo> privateKeysInfo = new ArrayList<>();
    Cursor cursor = context.getContentResolver().query(new KeysDaoSource().getBaseContentUri(), null, null, null, null);

    KeyStoreCryptoManager keyStoreCryptoManager = KeyStoreCryptoManager.getInstance(context);

    if (cursor != null && cursor.moveToFirst()) {
      do {
        String longId = cursor.getString(cursor.getColumnIndex(KeysDaoSource.COL_LONG_ID));
        String pubKey = cursor.getString(cursor.getColumnIndex(KeysDaoSource.COL_PUBLIC_KEY));

        String randomVector = KeyStoreCryptoManager.normalizeAlgorithmParameterSpecString(longId);

        String prvKey = keyStoreCryptoManager.decrypt(cursor.getString(
            cursor.getColumnIndex(KeysDaoSource.COL_PRIVATE_KEY)), randomVector);
        String passphrase = keyStoreCryptoManager.decrypt(cursor.getString(
            cursor.getColumnIndex(KeysDaoSource.COL_PASSPHRASE)), randomVector);

        PgpKeyInfo pgpKeyInfo = new PgpKeyInfo(longId, prvKey, pubKey);
        PrivateKeyInfo privateKeyInfo = new PrivateKeyInfo(pgpKeyInfo, passphrase);

        privateKeysInfo.add(privateKeyInfo);
      } while (cursor.moveToNext());
    }

    if (cursor != null) {
      cursor.close();
    }
    return privateKeysInfo;
  }

  /**
   * Check is backup of keys exist in the database.
   *
   * @return <tt>Boolean</tt> true if exists one or more private keys, false otherwise;
   */
  public static boolean hasBackup(Context context) {
    Cursor cursor = context.getContentResolver().query(new KeysDaoSource().getBaseContentUri(), null, null, null, null);

    boolean hasBackup = false;
    if (cursor != null && cursor.moveToFirst()) {
      hasBackup = cursor.getCount() > 0;
    }

    if (cursor != null) {
      cursor.close();
    }

    return hasBackup;
  }

  /**
   * Generate a new name for the private key which will be exported.
   *
   * @param email The user email.
   * @return A generated name for a new file.
   */
  public static String genPrivateKeyName(String email) {
    String sanitizedEmail = email.replaceAll("[^a-z0-9]", "");
    return "flowcrypt-backup-" + sanitizedEmail + ".key";
  }

  /**
   * Generate a private keys backup for the given account.
   *
   * @param context Interface to global information about an application environment.
   * @param account The given account
   * @return A string which includes private keys
   */
  public static String genPrivateKeysBackup(Context context, AccountDao account)
      throws PrivateKeyStrengthException, DifferentPassPhrasesException,
      NoPrivateKeysAvailableException, IOException, NodeException {
    StringBuilder builder = new StringBuilder();
    String email = account.getEmail();
    List<String> longIdsByEmail = new UserIdEmailsKeysDaoSource().getLongIdsByEmail(context, email);
    String[] longids = longIdsByEmail.toArray(new String[0]);
    KeysStorageImpl keysStorage = KeysStorageImpl.getInstance(context);
    List<PgpKeyInfo> pgpKeyInfoList = keysStorage.getFilteredPgpPrivateKeys(longids);

    if (CollectionUtils.isEmpty(pgpKeyInfoList)) {
      throw new NoPrivateKeysAvailableException(context, account.getEmail());
    }

    String firstPassPhrase = null;

    for (int i = 0; i < pgpKeyInfoList.size(); i++) {
      PgpKeyInfo pgpKeyInfo = pgpKeyInfoList.get(i);

      String passPhrase = keysStorage.getPassphrase(pgpKeyInfo.getLongid());

      if (i == 0) {
        firstPassPhrase = passPhrase;
      } else if (!passPhrase.equals(firstPassPhrase)) {
        throw new DifferentPassPhrasesException("The keys have different pass phrase");
      }

      if (TextUtils.isEmpty(passPhrase)) {
        throw new PrivateKeyStrengthException("Empty pass phrase");
      }

      Zxcvbn zxcvbn = new Zxcvbn();
      double measure = zxcvbn.measure(passPhrase, Arrays.asList(Constants.PASSWORD_WEAK_WORDS)).getGuesses();
      ZxcvbnStrengthBarResult passwordStrength = NodeCallsExecutor.zxcvbnStrengthBar(measure);

      if (passwordStrength != null) {
        switch (passwordStrength.getWord().getWord()) {
          case Constants.PASSWORD_QUALITY_WEAK:
          case Constants.PASSWORD_QUALITY_POOR:
            throw new PrivateKeyStrengthException("Pass phrase too weak");
        }
      }

      List<NodeKeyDetails> nodeKeyDetailsList = NodeCallsExecutor.parseKeys(pgpKeyInfo.getPrivate());
      NodeKeyDetails keyDetails = nodeKeyDetailsList.get(0);

      String encryptedKey;

      if (keyDetails.isDecrypted()) {
        EncryptKeyResult encryptKeyResult = NodeCallsExecutor.encryptKey(pgpKeyInfo.getPrivate(), passPhrase);
        encryptedKey = encryptKeyResult.getEncryptedKey();
      } else {
        encryptedKey = keyDetails.getPrivateKey();
      }

      builder.append(i > 0 ? "\n" + encryptedKey : encryptedKey);
    }

    return builder.toString();
  }

  /**
   * Get public keys for recipients + keys of the sender;
   *
   * @param context     Interface to global information about an application environment.
   * @param contacts    A list which contains recipients
   * @param account     The given account
   * @param senderEmail The sender email
   * @return A list of public keys.
   * @throws NoKeyAvailableException
   */
  public static List<String> getRecipientsPubKeys(Context context, List<String> contacts,
                                                  AccountDao account, String senderEmail)
      throws NoKeyAvailableException, IOException, NodeException {
    ArrayList<String> publicKeys = new ArrayList<>();
    List<PgpContact> pgpContacts = new ContactsDaoSource().getPgpContacts(context, contacts);

    for (PgpContact pgpContact : pgpContacts) {
      if (pgpContact != null && !TextUtils.isEmpty(pgpContact.getPubkey())) {
        publicKeys.add(pgpContact.getPubkey());
      }
    }

    publicKeys.add(getSenderPublicKey(context, account, senderEmail));

    return publicKeys;
  }

  /**
   * Get a public key of the sender;
   *
   * @param context     Interface to global information about an application environment.
   * @param account     The given account
   * @param senderEmail The sender email
   * @return <tt>String</tt> The sender public key.
   * @throws NoKeyAvailableException
   */
  public static String getSenderPublicKey(Context context, AccountDao account, String senderEmail) throws
      NoKeyAvailableException, IOException, NodeException {
    UserIdEmailsKeysDaoSource userIdEmailsKeysDaoSource = new UserIdEmailsKeysDaoSource();
    List<String> longIds = userIdEmailsKeysDaoSource.getLongIdsByEmail(context, senderEmail);

    if (longIds.isEmpty()) {
      if (account.getEmail().equalsIgnoreCase(senderEmail)) {
        throw new NoKeyAvailableException(context, account.getEmail(), null);
      } else {
        longIds = userIdEmailsKeysDaoSource.getLongIdsByEmail(context, account.getEmail());
        if (longIds.isEmpty()) {
          throw new NoKeyAvailableException(context, account.getEmail(), senderEmail);
        }
      }
    }

    PgpKeyInfo pgpKeyInfo = KeysStorageImpl.getInstance(context).getPgpPrivateKey(longIds.get(0));
    if (pgpKeyInfo != null) {
      List<NodeKeyDetails> details = NodeCallsExecutor.parseKeys(pgpKeyInfo.getPrivate());
      if (CollectionUtils.isEmpty(details)) {
        throw new IllegalStateException("There are no details about the given private key");
      }

      if (details.size() > 1) {
        throw new IllegalStateException("A wrong private key");
      }

      return details.get(0).getPublicKey();
    }

    throw new IllegalArgumentException("Internal error: PgpKeyInfo is null!");
  }
}

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/SecurityUtils.kt
/*
 * © 2016-present FlowCrypt a.s. Limitations apply. Contact human@flowcrypt.com
 * Contributors:
 *   DenBond7
 *   Ivan Pizhenko
 */

package com.flowcrypt.email.security

import android.content.Context
import com.flowcrypt.email.R
import com.flowcrypt.email.core.msg.RawBlockParser
import com.flowcrypt.email.database.FlowCryptRoomDatabase
import com.flowcrypt.email.database.entity.AccountEntity
import com.flowcrypt.email.extensions.org.bouncycastle.openpgp.armor
import com.flowcrypt.email.extensions.org.bouncycastle.openpgp.toPgpKeyDetails
import com.flowcrypt.email.extensions.org.pgpainless.key.info.usableForEncryption
import com.flowcrypt.email.security.model.PgpKeyDetails
import com.flowcrypt.email.security.pgp.PgpArmor
import com.flowcrypt.email.security.pgp.PgpKey
import com.flowcrypt.email.security.pgp.PgpPwd
import com.flowcrypt.email.util.exception.DifferentPassPhrasesException
import com.flowcrypt.email.util.exception.EmptyPassphraseException
import com.flowcrypt.email.util.exception.NoKeyAvailableException
import com.flowcrypt.email.util.exception.NoPrivateKeysAvailableException
import com.flowcrypt.email.util.exception.PrivateKeyStrengthException
import org.bouncycastle.bcpg.ArmoredOutputStream
import org.bouncycastle.openpgp.PGPPublicKeyRing
import org.pgpainless.key.OpenPgpV4Fingerprint
import org.pgpainless.key.info.KeyRingInfo
import org.pgpainless.util.Passphrase
import java.io.ByteArrayOutputStream
import java.io.OutputStream
import java.nio.charset.StandardCharsets

/**
 * This class help to receive security information.
 *
 * @author Denys Bondarenko
 */
class SecurityUtils {
  companion object {
    /**
     * Generate a new name for the private key which will be exported.
     *
     * @param email The user email.
     * @return A generated name for a new file.
     */
    @JvmStatic
    fun genPrivateKeyName(email: String): String {
      val sanitizedEmail = email.replace("[^a-z0-9]".toRegex(), "")
      return "flowcrypt-backup-$sanitizedEmail.key"
    }

    /**
     * Generate a private keys backup for the given account.
     *
     * @param context Interface to global information about an application environment.
     * @param account The given account
     * @return A string which includes private keys
     */
    fun genPrivateKeysBackup(context: Context, account: AccountEntity): String {
      val builder = StringBuilder()
      val keysStorage = KeysStorageImpl.getInstance(context.applicationContext)
      val keys = keysStorage.getPGPSecretKeyRings()

      if (keys.isEmpty()) {
        throw NoPrivateKeysAvailableException(context, account.email)
      }

      val fingerprintsWithEmptyPassphrase = keysStorage.getFingerprintsWithEmptyPassphrase()
      if (fingerprintsWithEmptyPassphrase.isNotEmpty()) {
        throw EmptyPassphraseException(
          fingerprints = fingerprintsWithEmptyPassphrase,
          message = context.getString(R.string.empty_pass_phrase)
        )
      }

      var firstPassPhrase: Passphrase? = null

      for (i in keys.indices) {
        val key = keys[i]
        val fingerprint = OpenPgpV4Fingerprint(key)
        val passPhrase = keysStorage.getPassphraseByFingerprint(fingerprint.toString())

        if (i == 0) {
          firstPassPhrase = passPhrase
        } else if (passPhrase != firstPassPhrase) {
          throw DifferentPassPhrasesException(context.getString(R.string.keys_have_different_pass_phrase))
        }

        if (passPhrase == null || passPhrase.isEmpty) {
          throw PrivateKeyStrengthException(context.getString(R.string.empty_pass_phrase))
        }

        PgpPwd.checkForWeakPassphrase(passPhrase)

        val keyDetails =
          key.toPgpKeyDetails(account.clientConfiguration?.shouldHideArmorMeta() ?: false)
        val encryptedKey = if (keyDetails.isFullyDecrypted) {
          PgpKey.encryptKey(keyDetails.privateKey ?: throw IllegalStateException(), passPhrase)
        } else {
          keyDetails.privateKey
        }

        builder.append(if (i > 0) "\n" + encryptedKey!! else encryptedKey)
      }

      return builder.toString()
    }

    /**
     * Get public keys for recipients;
     *
     * @param context     Interface to global information about an application environment.
     * @param emails      A list which contains recipients
     * @return A list of public keys.
     * @throws NoKeyAvailableException
     */
    @JvmStatic
    fun getRecipientsUsablePubKeys(context: Context, emails: MutableList<String>): List<String> {
      val publicKeys = mutableListOf<String>()
      val recipientsWithPubKeys = FlowCryptRoomDatabase.getDatabase(context).recipientDao()
        .getRecipientsWithPubKeysByEmails(emails)

      for (recipientWithPubKeys in recipientsWithPubKeys) {
        for (publicKeyEntity in recipientWithPubKeys.publicKeys) {
          val pgpKeyDetailsList = PgpKey.parseKeys(publicKeyEntity.publicKey).pgpKeyDetailsList
          for (pgpKeyDetails in pgpKeyDetailsList) {
            if (!pgpKeyDetails.isExpired && !pgpKeyDetails.isRevoked) {
              publicKeys.add(pgpKeyDetails.publicKey)
            }
          }
        }
      }

      return publicKeys
    }

    /**
     * Get sender public keys. We use this method to encrypt data for all user's available keys.
     *
     * @param context     Interface to global information about an application environment.
     * @param senderEmail The sender email
     */
    fun getSenderPublicKeys(
      context: Context,
      senderEmail: String
    ): List<String> {
      val keysStorage = KeysStorageImpl.getInstance(context.applicationContext)
      val matchingKeyRingInfoList = keysStorage.getPGPSecretKeyRingsByUserId(senderEmail)
        .map { KeyRingInfo(it) }.filter { it.usableForEncryption() }
      if (matchingKeyRingInfoList.isEmpty()) {
        throw IllegalStateException("There are no usable for encryption keys for $senderEmail")
      }
      return matchingKeyRingInfoList.map { PGPPublicKeyRing(it.publicKeys).armor() }
    }

    /**
     * Get a sender PGP keys details.
     *
     * @param context     Interface to global information about an application environment.
     * @param account     The given account
     * @param senderEmail The sender email
     * @throws NoKeyAvailableException
     */
    @JvmStatic
    fun getSenderPgpKeyDetails(
      context: Context,
      account: AccountEntity,
      senderEmail: String
    ): PgpKeyDetails {
      val keysStorage = KeysStorageImpl.getInstance(context.applicationContext)
      val pgpSecretKeyRing = keysStorage.getFirstUsableForEncryptionPGPSecretKeyRing(senderEmail)
        ?: if (account.email.equals(senderEmail, ignoreCase = true)) {
          throw NoKeyAvailableException(context, account.email)
        } else {
          throw NoKeyAvailableException(context, account.email, senderEmail)
        }

      return keysStorage.getPgpKeyDetailsList(listOf(pgpSecretKeyRing)).first()
    }

    /**
     * Filter an input file name to use only a safe part.
     */
    fun sanitizeFileName(originalFileName: String?): String {
      return originalFileName?.split("/")?.lastOrNull()?.replace("\\", "") ?: "unnamed"
    }

    /**
     * Check if the file extension fits the encrypted pattern.
     * If yes - it can mean the file is encrypted
     */
    fun isPossiblyEncryptedData(fileName: String?): Boolean {
      return RawBlockParser.ENCRYPTED_FILE_REGEX.containsMatchIn(fileName ?: "")
    }

    fun armor(
      hideArmorMeta: Boolean = false,
      headers: List<Pair<String, String>>? = PgpArmor.FLOWCRYPT_HEADERS,
      encode: (outputStream: OutputStream) -> Unit
    ): String {
      ByteArrayOutputStream().use { out ->
        ArmoredOutputStream(out).use { armoredOut ->
          if (headers != null) {
            if (hideArmorMeta) {
              armoredOut.clearHeaders()
            } else {
              for (header in headers) {
                armoredOut.setHeader(header.first, header.second)
              }
            }
          }
          encode.invoke(armoredOut)
        }
        return String(out.toByteArray(), StandardCharsets.US_ASCII)
      }
    }
  }
}
